### PR TITLE
Fix IN operator related issues in sql filter parser

### PR DESCRIPTION
**Pull request by WeichenXu123**
## Related Issues/PRs

#xxx

## What changes are proposed in this pull request?

The sqlparse<4.0.0 packages does not work with current mlflow sql filter parser IN operator: If sqlparse<4.0.0, the IN operator filter will be skipped.

So I add version restriction to be `sqlparse>=0.4.0` and remove useless code.

## How is this patch tested?

Unit tests.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Click the `Details` link on the `Preview docs` check.
2. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/search_utils.py [hunk 1]
```diff
@@ -1018,13 +980,5 @@ def _invalid_statement_token_search_model_registry(cls, token):
...
         elif token.match(ttype=TokenType.Keyword, values=["AND", "IN"]):
             return False
-        elif token.match(ttype=TokenType.Operator.Comparison, values=["IN"]):
-            # `IN` is a comparison token in sqlparse >= 0.4.0:
-            # https://github.com/andialbrecht/sqlparse/pull/567
-            if token.value == "IN":
-                # This case it represent the IN filter parsed failed.
-                # e.g. "run_id IN"
-                return True
-            return False
```
Comment (harupy) [inline mlflow/utils/search_utils.py:1028]: WHy do we need to remove this block?

File: mlflow/utils/search_utils.py [hunk 2]
```diff
@@ -677,47 +677,9 @@ def _parse_list_from_sql_token(cls, token):
                 error_code=INVALID_PARAMETER_VALUE,
             )
 
-    @classmethod
-    def _get_comparison_for_model_registry(cls, comparison, valid_search_keys):
```
Comment (harupy) [inline mlflow/utils/search_utils.py:681]: Why do we need to remove this method?

File: mlflow/utils/search_utils.py (hunk 1 above)
Comment (WeichenXu123) [reply mlflow/utils/search_utils.py:1028]: This is code for compatibility handling for sqlparse<4.0.0: 
In sqlparse<4.0.0, IN operator is not a `Comparison` type instance, but a Token type instance with `ttype=TokenType.Operator.Comparison, values=["IN"]`
but, current parsing logic already breaks the logic here under sqlparse<4.0.0 .

File: requirements/core-requirements.txt [hunk 3]
```diff
@@ -11,5 +11,6 @@ pandas<2
 prometheus-flask-exporter<1
 querystring_parser<2
 sqlalchemy<2,>=1.4.0
+sqlparse>=0.4.0
```
Comment (harupy) [inline requirements/core-requirements.txt:14]: See https://github.com/mlflow/mlflow/blob/80564db826e5a4ce9d1a18eee36d252b03e1ab6c/requirements/core-requirements.txt#L1-L3

File: mlflow/utils/search_utils.py (hunk 2 above)
Comment (WeichenXu123) [reply mlflow/utils/search_utils.py:681]: The method is not used any more.

Comment (harupy) [reply mlflow/utils/search_utils.py:681]: If so, please add a comment or file a separate PR.

Comment (harupy) [reply mlflow/utils/search_utils.py:681]: Or you can file a different PR. `_get_comparison_for_model_registry` is unrelated to sqlparse's issue.

Comment (WeichenXu123) [reply mlflow/utils/search_utils.py]: File a new PR https://github.com/mlflow/mlflow/pull/6412 for this.

File: mlflow/utils/search_utils.py (hunk 1 above)
Comment (harupy) [reply mlflow/utils/search_utils.py:1028]: This code is for compatibility for `sqlparse>=0.4.0` as the comment explains.

Comment (WeichenXu123) [reply mlflow/utils/search_utils.py:1028]: No. The comment is not clear.
Using `sqlparse>=0.4.0`, this block code will never be run.

**Comment by harupy** [comment]
```
# Check sqlparse<0.4.0 is installed
> pip list | grep sqlparse
sqlparse                             0.3.1

# Print a comparison item
> gd
diff --git a/mlflow/utils/search_utils.py b/mlflow/utils/search_utils.py
index f77a596ed..5afb51c81 100644
--- a/mlflow/utils/search_utils.py
+++ b/mlflow/utils/search_utils.py
@@ -730,6 +730,7 @@ class SearchUtils:
             "comparator": stripped_comparison[1].value,
             "value": value,
         }
+        print(comp)
         return comp
 
     @classmethod

# Run tests
> pytest tests/store/model_registry/test_sqlalchemy_store.py::TestSqlAlchemyStoreSqlite::test_search_model_versions -s
========================================================================================================================================================== test session starts ===========================================================================================================================================================
platform linux -- Python 3.7.10, pytest-6.2.5, py-1.10.0, pluggy-1.0.0 -- /home/haru/miniconda3/envs/mlflow-dev-env/bin/python
cachedir: .pytest_cache
rootdir: /home/haru/Desktop/repositories/mlflow, configfile: pytest.ini
plugins: cov-3.0.0, localserver-0.6.0
collected 1 item                                                                                                                                                                                                                                                                                                                         

tests/store/model_registry/test_sqlalchemy_store.py::TestSqlAlchemyStoreSqlite::test_search_model_versions {'key': 'name', 'comparator': '=', 'value': 'test_for_search_MV'}
{'key': 'run_id', 'comparator': '=', 'value': '10e5bd10eabf4f56b0b56d4e206ecf59'}
{'key': 'run_id', 'comparator': '=', 'value': 'a93b9d74be2b474d8cf214b79f6f4de8'}
👇👇👇
{'key': 'run_id', 'comparator': 'IN', 'value': ('10e5bd10eabf4f56b0b56d4e206ecf59', 'a93b9d74be2b474d8cf214b79f6f4de8')}
{'key': 'source_path', 'comparator': '=', 'value': 'A/D'}
{'key': 'source_path', 'comparator': '=', 'value': 'A'}
{'key': 'source_path', 'comparator': '=', 'value': 'A/'}
{'key': 'source_path', 'comparator': '=', 'value': ''}
{'key': 'name', 'comparator': '=', 'value': 'test_for_search_MV'}
{'key': 'source_path', 'comparator': '=', 'value': 'A/D'}
{'key': 'run_id', 'comparator': '=', 'value': '10e5bd10eabf4f56b0b56d4e206ecf59'}
PASSED
```

**Comment by harupy** [comment]
Prior to #6320, `_process_statement_tokens` was used when parsing the filter string:

https://github.com/mlflow/mlflow/blob/80564db826e5a4ce9d1a18eee36d252b03e1ab6c/mlflow/utils/search_utils.py#L722-L734

**Comment by WeichenXu123** [comment]
> It looks like a filter string containing `IN` was parsed correctly prior to #6320.
> 
> ```
> # Check sqlparse<0.4.0 is installed
> > pip list | grep sqlparse
> sqlparse                             0.3.1
> 
> # Print a comparison item
> > gd
> diff --git a/mlflow/utils/search_utils.py b/mlflow/utils/search_utils.py
> index f77a596ed..5afb51c81 100644
> --- a/mlflow/utils/search_utils.py
> +++ b/mlflow/utils/search_utils.py
> @@ -730,6 +730,7 @@ class SearchUtils:
>              "comparator": stripped_comparison[1].value,
>              "value": value,
>          }
> +        print(comp)
>          return comp
>  
>      @classmethod
> 
> # Run tests
> > pytest tests/store/model_registry/test_sqlalchemy_store.py::TestSqlAlchemyStoreSqlite::test_search_model_versions -s
> ========================================================================================================================================================== test session starts ===========================================================================================================================================================
> platform linux -- Python 3.7.10, pytest-6.2.5, py-1.10.0, pluggy-1.0.0 -- /home/haru/miniconda3/envs/mlflow-dev-env/bin/python
> cachedir: .pytest_cache
> rootdir: /home/haru/Desktop/repositories/mlflow, configfile: pytest.ini
> plugins: cov-3.0.0, localserver-0.6.0
> collected 1 item                                                                                                                                                                                                                                                                                                                         
> 
> tests/store/model_registry/test_sqlalchemy_store.py::TestSqlAlchemyStoreSqlite::test_search_model_versions {'key': 'name', 'comparator': '=', 'value': 'test_for_search_MV'}
> {'key': 'run_id', 'comparator': '=', 'value': '10e5bd10eabf4f56b0b56d4e206ecf59'}
> {'key': 'run_id', 'comparator': '=', 'value': 'a93b9d74be2b474d8cf214b79f6f4de8'}
> 👇👇👇
> {'key': 'run_id', 'comparator': 'IN', 'value': ('10e5bd10eabf4f56b0b56d4e206ecf59', 'a93b9d74be2b474d8cf214b79f6f4de8')}
> {'key': 'source_path', 'comparator': '=', 'value': 'A/D'}
> {'key': 'source_path', 'comparator': '=', 'value': 'A'}
> {'key': 'source_path', 'comparator': '=', 'value': 'A/'}
> {'key': 'source_path', 'comparator': '=', 'value': ''}
> {'key': 'name', 'comparator': '=', 'value': 'test_for_search_MV'}
> {'key': 'source_path', 'comparator': '=', 'value': 'A/D'}
> {'key': 'run_id', 'comparator': '=', 'value': '10e5bd10eabf4f56b0b56d4e206ecf59'}
> PASSED
> ```

Yes.
Previous code using sqlparse<0.4.0 also works fine for "IN" operator, because previous code it uses another implementation https://github.com/mlflow/mlflow/blob/f3199e2c9354b49f30dc9325c144d028ac170e03/mlflow/utils/search_utils.py#L772 
Note previous implementation is hardcoded for only one filter.

File: mlflow/utils/search_utils.py (hunk 1 above)
Comment (harupy) [reply mlflow/utils/search_utils.py:1028]: Test:

```diff
diff --git a/mlflow/utils/search_utils.py b/mlflow/utils/search_utils.py
index 2a3c5dbe6..f69cc5498 100644
--- a/mlflow/utils/search_utils.py
+++ b/mlflow/utils/search_utils.py
@@ -1019,6 +1019,7 @@ class SearchModelUtils(SearchUtils):
         elif token.match(ttype=TokenType.Keyword, values=["AND", "IN"]):
             return False
         elif token.match(ttype=TokenType.Operator.Comparison, values=["IN"]):
+            raise Exception(sqlparse.__version__)
             # `IN` is a comparison token in sqlparse >= 0.4.0:
             # https://github.com/andialbrecht/sqlparse/pull/567
             if token.value == "IN":
```

```
> pytest tests/store/model_registry/test_sqlalchemy_store.py::TestSqlAlchemyStoreSqlite::test_search_model_versions -s

...
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    @classmethod
    def _invalid_statement_token_search_model_registry(cls, token):
        if isinstance(token, (Comparison, Identifier, Parenthesis)):
            return False
        elif token.is_whitespace:
            return False
        elif token.match(ttype=TokenType.Keyword, values=["AND", "IN"]):
            return False
        elif token.match(ttype=TokenType.Operator.Comparison, values=["IN"]):
>           raise Exception(sqlparse.__version__)
E           Exception: 0.4.2


mlflow/utils/search_utils.py:1022: Exception
```

Comment (WeichenXu123) [reply mlflow/utils/search_utils.py:1028]: You misread above comment, I said `this block code will never be run except invalid filter such as run_id IN`,
Your test is the case `invalid filter such as run_id IN`

Comment (harupy) [reply mlflow/utils/search_utils.py:1028]: You're right. I thought the error was thrown in a test case with a valid filter string, but it actually wasn't.

**Comment by harupy** [comment]
We could use a dirty trick like this to handle the `IN` comparison in `sqlparse < 0.4.0`, but it's not worth the effort:

```diff
diff --git a/mlflow/utils/search_utils.py b/mlflow/utils/search_utils.py
index 2a3c5dbe6..eddd5af64 100644
--- a/mlflow/utils/search_utils.py
+++ b/mlflow/utils/search_utils.py
@@ -4,6 +4,7 @@ import operator
 import re
 import ast
 import shlex 
 import sqlparse
 from sqlparse.sql import (
@@ -936,6 +937,29 @@ class SearchModelUtils(SearchUtils):
             raise MlflowException.invalid_parameter_value(
                 "Invalid clause(s) in filter string: %s" % invalid_clauses
             )
+        counter = 0
+        tokens = [t for t in statement.tokens if not t.is_whitespace]
+        new_tokens = []
+        while counter < len(tokens):
+            print(tokens[counter].__class__)
+            token = tokens[counter]
+            if (
+                counter + 2 < len(tokens)
+                and not isinstance(tokens[counter], Comparison)
+                and isinstance(tokens[counter], Identifier)
+                and tokens[counter].value == "run_id"
+                and tokens[counter + 1].match(ttype=TokenType.Keyword, values=["IN"])
+                and isinstance(tokens[counter + 2], Parenthesis)
+            ):
+                new_tokens.append(Comparison(TokenList(tokens[counter : counter + 3])))
+                counter += 3
+            else:
+                new_tokens.append(token)
+                counter += 1
+
+        if any(map(lambda t: not isinstance(t, (Comparison, Token)), new_tokens)):
+            raise MlflowException("Invalid clause(s) in filter string", INVALID_PARAMETER_VALUE)
+
         return [cls._get_comparison(t) for t in statement.tokens if isinstance(t, Comparison)]
 
     @classmethod
```

File: requirements/core-requirements.yaml [hunk 4]
```diff
@@ -41,6 +41,11 @@ sqlalchemy:
   minimum: "1.4.0"
   max_major_version: 1
 
+sqlparse:
+  pip_release: sqlparse
+  minimum: "0.4.0"
+  max_major_version: 1
```
Comment (harupy) [inline requirements/core-requirements.yaml:47]:
```suggestion
  max_major_version: 0
```

**Review by harupy — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/6411/files#r938783289 is addressed.

File: requirements/core-requirements.yaml [hunk 5]
```diff
@@ -41,6 +41,11 @@ sqlalchemy:
   minimum: "1.4.0"
   max_major_version: 1
 
+sqlparse:
+  pip_release: sqlparse
+  minimum: "0.4.0"
+  max_major_version: 0
```
Comment (dbczumar) [inline requirements/core-requirements.yaml:47]: sqlparse is already present in `skinny-requirements`. Shall we bump the minimum version to `0.4.0` there?

Comment (dbczumar) [reply requirements/core-requirements.yaml:47]: https://github.com/mlflow/mlflow/blob/bfbef7bfb9901da279b6ed7929831d02c4070fc4/requirements/skinny-requirements.yaml#L62-L66

**Review by dbczumar — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/6411#discussion_r940561669 is addressed. Thanks @WeichenXu123 !

File: requirements/core-requirements.yaml (hunk 5 above)
Comment (WeichenXu123) [reply requirements/core-requirements.yaml]: addressed.

**Pull request merged**